File: 回溯算法/05电话号码组合.py
# 17. 电话号码的字母组合

def letterCombinations(digits: str):
    if digits == "":
        return []
    
    nums_dict = {2: 'abc', 3: 'def', 4: 'ghi', 5: 'jkl', 6: 'mno',
                    7: 'pqrs', 8: 'tuv', 9: 'wxyz'}
    
    digits_list = list(map(int, [i for i in digits]))

    res = []
    track = []
    di_length = len(digits_list)
    def backtrack(idx, track):
        if len(track) == di_length:
            res.append(''.join(track))
            return 

        n = digits_list[idx]  # 数字
        c = nums_dict[n]  # 字符集
    
        for i in range(len(c)):
            track.append(c[i])
            backtrack(idx+1, track)
            track.pop(-1)
    backtrack(0, [])
    return res

File: 回溯算法/05电话号码组合.java
class Solution {
    List<String> res = new ArrayList<>();
    StringBuffer temp = new StringBuffer();

    /**
     * 17. 电话号码的字母组合
     * @param digits
     * @return
     */
    public List<String> letterCombinations(String digits) {
        if (digits.length() < 1)
            return res;
        String[] numString = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        drawBack(digits, numString, 0);
        return res;

    }

    private void drawBack(String digits, String[] numString, int num) {
        if (num == digits.length()) {
            res.add(temp.toString());
            return;
        }
        int n = digits.charAt(num) - '0';
        String str = numString[n];

        for (int i = 0; i < str.length(); i++) {
            // 选择
            temp.append(str.charAt(i));
            // 下一个数字的字母组合
            // 注意此处是 num+1
            drawBack(digits, numString, num+1);
            // 回撤选择
            temp.deleteCharAt(temp.length() - 1);
        }
    }
}